math/geometry.h

```diff
@@ -253,6 +262,26 @@ void ProjectPointOntoLineSegment(const Eigen::Matrix<T, 2, 1>& point,
   *squared_distance = (point - (*projected_point)).squaredNorm();
 }
 
+// Project a given point onto a line segment.
+// The line segment is defined by its two vertices vertex_a and vertex_b
+// Returns the location of the projection.
+template <typename T>
+Eigen::Matrix<T, 2, 1> ProjectPointOntoLineSegment(
+    const Eigen::Matrix<T, 2, 1>& point,
+    const Eigen::Matrix<T, 2, 1>& vertex_a,
+    const Eigen::Matrix<T, 2, 1>& vertex_b) {
+  Eigen::Matrix<T, 2, 1> line_segment = vertex_b - vertex_a;
+  Eigen::Matrix<T, 2, 1> point_vector = point - vertex_a;
+
+  // Project the point onto the line segment, capping it at the two end points
+  float scalar_projection =
+      point_vector.dot(line_segment) / line_segment.squaredNorm();
+
+  scalar_projection = fmax(0, fmin(1, scalar_projection));
```

Comment:
Instead of using `fmax` this should use `std::max<T>` to play well with non-POD types, e.g. for AD with Ceres.

Reply:
Same for `fmax`

Reply:
Done.